python/ray/_private/log_rotation_with_pipe.py

```diff
@@ -0,0 +1,142 @@
+        return False
+
+    def writable(self) -> bool:
+        return True
+
+    def seekable(self) -> bool:
+        return False
+
+
+def open_pipe_with_file_handlers(log_name: str, handlers) -> IO[AnyStr]:
+    """Stream content into pipe, which will be listened and dumped to provided
+    handlers."""
+    read_fd, write_fd = os.pipe()
+
+    log_content = []
+    stopped = threading.Event()
+    cond = threading.Condition()
+
+    # Have to use different logger names for different sinks.
+    logger = logging.getLogger(f"logger-{log_name}")
+    logger.setLevel(logging.INFO)
+    for handler in handlers:
+        logger.addHandler(handler)
+
+    # Setup read thread, which continuous read content out of pipe and append to buffer.
+    def read_log_content_from_pipe():
+        with os.fdopen(read_fd, "r") as pipe_reader:
+            while True:
+                line = pipe_reader.readline()
+
+                with cond:
+                    # An empty line is read over, which indicates write side has
+                    # closed.
+                    if line == _EOF_TOKEN:
+                        stopped.set()
+                        cond.notify()
+                        return
+
+                    # Only buffer new line when not empty.
+                    stripped = line.strip()
+                    if not stripped:
+                        continue
+
+                    # Append content and continue.
+                    log_content.append(stripped)
+                    cond.notify()
+
+    # Setup logging thread, which continues read content out of log buffer and persist
+    # via logger.
+    # Two threads are used here to make sure continuous read from pipe thus avoid
+    # blocking write from application.
+    def dump_log_content_to_buffer():
+        # Current line to log.
+        cur_content = None
+
+        while True:
+            with cond:
+                while not log_content and not stopped.is_set():
+                    cond.wait()
+
+                if log_content:
+                    cur_content = log_content.pop(0)
+                else:
+                    assert stopped.is_set()
+                    return
+
+            # Log and perform IO operation out of critical section.
+            logger.info(cur_content)
+
+    # Python `atexit` hook only invokes when no other non-daemon threads running.
+    read_thread = threading.Thread(target=read_log_content_from_pipe, daemon=True)
+    dump_thread = threading.Thread(target=dump_log_content_to_buffer, daemon=True)
+    read_thread.start()
+    dump_thread.start()
+
+    pipe_write_stream = PipeStreamWriteHandle(write_fd, log_name=log_name)
+
+    def cleanup():
+        pipe_write_stream.write(_EOF_TOKEN)
+        pipe_write_stream.flush()
+        pipe_write_stream.close()
+
+        # Synchronize to make sure everything flushed.
+        read_thread.join()
+        dump_thread.join()
+
+    atexit.register(cleanup)
+
+    return pipe_write_stream
```

Comment:
Do we need the `PipeStreamWriteHandle` wrapper, can we directly return `open(write_fd)`?

Reply:
`open` takes a path-like object, not a fd
Ref: https://docs.python.org/3/library/functions.html#open

Reply:
`fdopen` is the interface we want to wrap fd into a file object
Ref: https://docs.python.org/3/library/os.html#os.fdopen

Reply:
Can we use `fdopen` then?

Reply:
How to you plan to provide `name` and `write` then?

Reply:
I see. `write()` I think is supported by the object returned by `fdopen()`? The issue is `name` then. I guess you can do

```
file_like_io_object = os.fdopen(write_fd)
file_like_io_object.name = "xxxx"
```

Reply:
Just tried
```
Traceback (most recent call last):
  File "logging_redirection_test.py", line 13, in <module>
    _LOG_STDOUT_FNAME, _ROTATION_MAX_BYTE, _ROTATION_FILE_NUM)
  File "/home/ubuntu/py-logging-redirection/pipe_logging.py", line 90, in open_pipe_with_rotation
    pipe_write_stream.name = fname
AttributeError: attribute 'name' of '_io.TextIOWrapper' objects is not writable
```

Reply:
Code snippet:
```
pipe_write_stream = os.fdopen(write_fd, "w")
pipe_write_stream.name = fname
```

Reply:
Seems we can do https://stackoverflow.com/questions/60622854/how-to-instantiate-an-io-textiowrapper-object-with-a-name-attribute so we don't need to implement our own `IO` subclass with all these methods.